keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2469 @@
+generic plugin would be able to filter out nodes based on already allocated
+resources. But because it is stateless, it would not handle the use count
+restrictions correctly when multiple pods are pending and reference the same
+resource.
+
+### kubelet
+
+#### Managing resources
+
+kubelet must ensure that resources are ready for use on the node before running
+the first Pod that uses a specific resource instance and make the resource
+available elsewhere again when the last Pod has terminated that uses it. For
+both operations, kubelet calls a resource kubelet plugin as explained in the next
+section.
+
+Pods that are not listed in ReservedFor or where the ResourceClaim doesn't
+exist at all must not be allowed to run. Instead, a suitable event must be
+emitted which explains the problem. Such a situation can occur as part of
+downgrade scenarios.
+
+If this was the last Pod on the node that uses the specific
+resource instance, then NodeUnprepareResource (see below) must have been called
+successfully before allowing the pod to be deleted. This ensures that network-attached resource are available again
+for other Pods, including those that might get scheduled to other nodes. It
+also signals that it is safe to deallocate and delete the ResourceClaim.
+
+
+![kubelet](./kubelet.png)
+
+#### Communication between kubelet and resource kubelet plugin
+
+Resource kubelet plugins are discovered through the [kubelet plugin registration
+mechanism](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/#device-plugin-registration). A
+new "ResourcePlugin" type will be used in the Type field of the
+[PluginInfo](https://pkg.go.dev/k8s.io/kubelet/pkg/apis/pluginregistration/v1#PluginInfo)
+response to distinguish the plugin from device and CSI plugins.
+
+Under the advertised Unix Domain socket the kubelet plugin provides the following
+gRPC interface. It was inspired by
+[CSI](https://github.com/container-storage-interface/spec/blob/master/spec.md),
+with “volume” replaced by “resource” and volume specific parts removed.
+
+##### `NodePrepareResource`
+
+This RPC is called by kubelet when a Pod that wants to use the
+specified resource is scheduled on a node.  The Plugin SHALL assume
+that this RPC will be executed on the node where the resource will be
+used.  ResourceClaim.meta.Namespace, ResourceClaim.meta.UID,
+ResourceClaim.Name, ResourceClaim.meta.Namespace and
+ResourceClaimStatus.AllocationResult should be passed to the Plugin
+as parameters to identify the claim and perform resource preparation.
+
+ResourceClaim parameters(namespace, UUID, name) are useful for debugging
+and enable the resource driver to retrieve the full ResourceClaim object,
+should that ever be needed (normally it shouldn't).
+
+The Plugin SHALL return fully qualified device name[s].
+
+The Plugin SHALL ensure that there are json file[s] in CDI format
+for the allocated resource. These files SHALL be used by runtime to
```

Comment:
How are these files configured for the runtime? I assume plugins are containers so we need to dynamically pick up new CDI files?

Reply:
Plugin can either use existing files or create them and return fully qualified device names to the Kubelet. Kubelet passes these names down to a runtime through CRI. Runtime updates container spec using CDI file before creating a container, giving a container access to the devices, environment variables, mounts and hooks specified in the CDI file.

Reply:
One minor issues is standardization of directores for CDI files across runtimes but that problem is similar to CNI. If runtimes stick to standard directories than plugins don't need runtime specific paths. Also, runtimes can inotify watch and dynamically update CDI cache.

Reply:
I believe it has been standardized and the Go code in the cdi repo has code for accessing the files that is shared between runtimes.